### PR TITLE
fortran backend: implieddoLoop support added

### LINKED ISSUE #4113: enh: add support for ImpliedDoLoop in fortran backend
https://github.com/lfortran/lfortran/pull/4101#discussion_r1618927163

### PR DESCRIPTION
**Pull request by captainhaddock18**
Fixes #4113 . Please have a look.

### CODE REVIEW

File: src/libasr/codegen/asr_to_fortran.cpp [hunk 1]
```diff
@@ -1451,10 +1451,33 @@ class ASRToFortranVisitor : public ASR::BaseVisitor<ASRToFortranVisitor>
     // void visit_UnionConstructor(const ASR::UnionConstructor_t &x) {}
 
     void visit_ImpliedDoLoop(const ASR::ImpliedDoLoop_t &x) {
```
Comment (gxyd) [inline src/libasr/codegen/asr_to_fortran.cpp:1453]: We'll need to add a test for this, one way to test the support for `ImpliedDoLoop` support here is by adding `fortran` label for a test case in `integration_tests/CMakeLists.txt` e.g.;
```
RUN(NAME min_01 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc fortran)
```
the above line has `fortran` label in it, which indicates that the fortran backend (i.e. ASR -> Fortran) works for this test case and produces valid Fortran code.

We'll need to do the same for your changes as well, i.e. add `fortran` label for an existing test case (ideally) or a new test case, either way, it should indicate to us that the test-case used to fail earlier and now it passes with this PR.

File: src/libasr/codegen/asr_to_fortran.cpp [hunk 2]
```diff
@@ -1451,10 +1451,33 @@ class ASRToFortranVisitor : public ASR::BaseVisitor<ASRToFortranVisitor>
...
-        if (x.m_value) {
-            visit_expr(*x.m_value);
-            return;
-        }
+    std::string r = "(";
```
Comment (gxyd) [inline src/libasr/codegen/asr_to_fortran.cpp:1454]: Also, can we please format the code in better way by indenting it correctly?

**Review by gxyd — COMMENTED** [review]
Good start, if we can please address the above comments it would be nice.

File: src/libasr/codegen/asr_to_fortran.cpp (hunk 1 above)
Comment (HarshitaKalani) [reply src/libasr/codegen/asr_to_fortran.cpp:1453]: Yes.
Maybe you can add the fortran label for the implied do loop tests.
```diff
- RUN(NAME implied_do_loops1 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
+ RUN(NAME implied_do_loops1 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc fortran)
```
and similarly for the others
```
RUN(NAME implied_do_loops2 LABELS gfortran)
RUN(NAME implied_do_loops3 LABELS gfortran)
RUN(NAME implied_do_loops4 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
RUN(NAME implied_do_loops5 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
.
.
.
RUN(NAME data_implied_do_01 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
RUN(NAME data_implied_do_02 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
RUN(NAME data_implied_do_03 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
RUN(NAME data_implied_do_04 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
```

File: src/libasr/codegen/asr_to_fortran.cpp (hunk 2 above)
Comment (captainhaddock18) [reply src/libasr/codegen/asr_to_fortran.cpp:1454]: got it, I'll will make these changes, Thank you!

**Review by Pranavchiku — APPROVED** [review]
looks good to me.

**Comment by Pranavchiku** [comment]
```fortran
program main
integer :: i
print *, (i, i = 1, 10)
end program
```

```console
% lfortran a.f90 --show-fortran
program main
implicit none
integer(4) :: i
print *, (i, i = 1, 10)
end program main
```

**Review by gxyd — APPROVED** [review]
Looks good to me, thanks @captainhaddock18 for the work here.

**Comment by gxyd** [comment]
I'm squashing the commit and merging them, as the commit messages aren't very useful, if you can improve on writing the commit messages, that would make it easy to merge the PR directly without squashing the commits going forward, for now it's allright.

**Comment by Pranavchiku** [comment]
@captainhaddock18 you may take a look at `conventional commits`, will help you to write meaningful commits.

**Pull request merged**